packages/styled-components/src/sheet/Rehydration.js

```diff
@@ -67,10 +65,12 @@ const rehydrateSheetFromTag = (sheet: Sheet, style: HTMLStyleElement) => {
 
   for (let i = 0, l = cssRules.length; i < l; i++) {
     const cssRule = (cssRules[i]: any);
-
     if (typeof cssRule.cssText !== 'string') {
       // Avoid IE11 quirk where cssText is inaccessible on some invalid rules
       continue;
+    } else if (cssRule.type === IMPORT_RULE_TYPE) {
```

Comment:
Let’s revert this addition to avoid confusion later on?

Since we don’t have the new hoisting yet, we can’t just leave these out. Since SSR will aggressively avoid insertions, we’d likely remove these forever if we skip them during rehydration.

So let’s remove this change and we’ll figure changes here out once the SSR hoisting is implemented properly :)

Reply:
What did you have in mind left to do with the SSR hoisting? I adjusted some things in this PR per comments on your PR (which this will merge into). With the adjustments to the hoisting in your PR we do have the imports all hoisted to the top of the style tag, just with no markers since I couldn't find a marker that wouldn't result in something ending up invalid (`@import` rules are imported aggressively so the conditional media rule stuff didn't work out, and no other rules can be intermingled with `@import` and let following `@import` statements be valid). 

So, the flow that works out with ignoring this (from what I understand/tested):
1. `@import` rules are hoisted in the SSR output (see code changes above)
1. `@import` rules are ignored on rehydration since we can't effectively figure out what group they belong to. All other rules are still rehydrated normally
1. `@import` rules would be added (in a hoisted position) during a re-render/whenever the client-side logic takes over the tag

Does that sound right? I'd love to re-hydrate them instead, but I couldn't find an effective way to do that so if you have ideas I'm game! :)

Reply:
We do have them hoisted in the output.
But if we ignore them they will be removed and actually _won't be readded_ since the sheet has the `names` cache, which will ignore entire rule blocks when they've already been inserted.

This is important, because otherwise we'd readd rules on initial rendering/hydration, which is rather slow. Now, we could put in an exception that doesn't cache any rules during SSR with the name-cache that contain `@import` rules, but I'd suspect that this will potentially sometimes cause a flash on the page, as the `@import` is removed and then readded

Reply:
Ah, I'll have to look at that caching then since from what I could tell on my testing they're being re-added presently.

Reply:
That's likely because the tests never contain a full flow (at least not the `Rehydration.test.js`) of adding rules with names then rehydrating, and then readding the rules.

Instead the name cache is assumed to "just work" and not have any special properties apart from preventing components from adding a batch of rules twice.

Reply:
Sounds good! If you get time today to look at it, I can manage updating this if you update your PR (unless you decide to push here... Which I think you can?).

I'll look at adding a full flow test too, since that sounds handy to have 🙂

Reply:
Also note that I've been testing this using the sandbox app in the repo, but I haven't committed any changes to that since I wasn't sure if that generally stayed untouched 🙂 Before pushing the adjustments to your logic/PR I also pulled it into a sample app I have at work to make sure those fonts loaded without a flash (and they did), since I'm looking at v4 as well to try and fix the flash 🙂